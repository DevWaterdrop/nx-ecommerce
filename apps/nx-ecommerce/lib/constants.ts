export const LOVELY_ERROR = 'I HATE TYPESCRIPT+STRAPI+GRAPHQL';
export const MAXIMUM_IN_CART = 16;

export enum PRICE_FILTER {
  ZERO_TWENTY = 'zt',
  TWENTY_FORTY = 'tf',
  FORTY_EIGHTY = 'fe',
  EIGHTY_PLUS = 'ep',
}

export enum SORT_FILTER {
  PRICE_LOW_HIGH = 'plh',
  PRICE_HIGH_LOW = 'phl',
  NAME = 'name',
}

export const PRICES = [
  { value: PRICE_FILTER.ZERO_TWENTY, label: '0.00$ - 19.99$' },
  { value: PRICE_FILTER.TWENTY_FORTY, label: '20.00$ - 39.99$' },
  { value: PRICE_FILTER.FORTY_EIGHTY, label: '40.00$ - 79.99$' },
  { value: PRICE_FILTER.EIGHTY_PLUS, label: '80.00$+' },
];

export const SORT = [
  { value: SORT_FILTER.PRICE_LOW_HIGH, label: 'Price: low to high' },
  { value: SORT_FILTER.PRICE_HIGH_LOW, label: 'Price: high to low' },
  { value: SORT_FILTER.NAME, label: 'Name' },
];
