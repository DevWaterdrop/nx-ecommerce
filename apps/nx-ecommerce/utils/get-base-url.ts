export function getBaseURL(path = '') {
  return `${
    (process.env['NX_BASE_URL'] as string) || 'http://localhost:1337'
  }${path}`;
}
