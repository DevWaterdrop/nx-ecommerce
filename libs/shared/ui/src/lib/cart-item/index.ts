import { CartItem, CartItemProps } from './cart-item';

export { CartItem };
export type { CartItemProps };
