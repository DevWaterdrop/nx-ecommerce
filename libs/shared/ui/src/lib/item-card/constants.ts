import { ItemCardProps } from './item-card';
import flowerJPEG from '../../props/flower.jpeg';
import roomJPEG from '../../props/room.jpeg';

const ITEM: Required<ItemCardProps['item']> = {
  price: 65,
  name: 'Lovely flower',
  id: '#',
  images: [flowerJPEG, roomJPEG],
  smallDescription: 'For you, for me, for us',
  tag: 'new',
};

export const DEFAULT_PROPS: ItemCardProps = {
  item: ITEM,
  size: 'base',
  handleClick: () => console.log('clicked'),
};
