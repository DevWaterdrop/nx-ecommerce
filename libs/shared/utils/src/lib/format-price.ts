export function formatPrice(price: number) {
  return `${Math.max(price, 0).toFixed(2)}$`;
}
