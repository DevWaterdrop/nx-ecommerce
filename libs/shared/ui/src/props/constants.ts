import { ButtonProps } from '../lib/button';
import { CartItemProps } from '../lib/cart-item';
import { CartSidebarProps } from '../lib/cart-sidebar';
import { CategoryCardProps } from '../lib/category-card';
import { IconProps } from '../lib/icon/icon';
import { ItemCardProps } from '../lib/item-card';
import { NavButtonProps } from '../lib/nav-button';
import { OrdersTableProps } from '../lib/orders-table';
import { ProductGalleryProps } from '../lib/product-gallery';
import { ProductInformationProps } from '../lib/product-information';
import { ProductSidebarProps } from '../lib/product-sidebar';
import { SearchbarProps } from '../lib/searchbar';
import { SearchbarItemProps } from '../lib/searchbar-item';
import { SelectProps } from '../lib/select';
import flowerJPEG from './flower.jpeg';
import roomJPEG from './room.jpeg';

export const ITEM: ItemCardProps['item'] = {
  slug: 'lovely-flower',
  price: 65,
  name: 'Lovely flower',
  images: {
    data: [
      {
        attributes: {
          url: flowerJPEG,
        },
      },
      {
        attributes: {
          url: roomJPEG,
        },
      },
    ],
  },
  smallDescription: 'For you, for me, for us',
  description: 'Lovely description',
};

export const ITEM_SECOND: ItemCardProps['item'] = {
  slug: 'roomy',
  price: 50,
  name: 'Roomy',
  images: {
    data: [
      {
        attributes: {
          url: flowerJPEG,
        },
      },
      {
        attributes: {
          url: roomJPEG,
        },
      },
    ],
  },
  smallDescription: 'Hmmm, nice vase',
  description: 'Vaaaase',
};

export const ITEM_CARD_DEFAULT_PROPS: ItemCardProps = {
  item: ITEM,
  size: 'base',
  handleCartClick: () => console.log('cart clicked'),
  handleAmountCartClick: () => console.log('amount cart clicked'),
  handleFavoriteClick: () => console.log('favorite clicked'),
};

export const CATEGORY_CARD_DEFAULT_PROPS: CategoryCardProps = {
  category: {
    smallDescription:
      'Lorem ipsum dolor sit amet, consectetur adipisicing elit.',
    image: {
      data: {
        attributes: {
          url: roomJPEG,
        },
      },
    },
    slug: '1',
  },
};

export const NAV_BUTTON_DEFAULT_PROPS: NavButtonProps = {
  type: 'heart',
};

export const ICON_DEFAULT_PROPS: IconProps = {
  icon: 'cart',
};

export const BUTTON_DEFAULT_PROPS: ButtonProps = {
  children: 'hello',
  tag: 'button',
  size: 'base',
  type: 'primary',
  round: 'full',
};

export const SEARCHBAR_ITEM: SearchbarProps['items'][number] = {
  id: '111',
  attributes: {
    slug: 'lovely-flower',
    smallDescription: 'For you, for me, for us',
    name: 'Lovely flower',
    images: {
      data: [
        {
          attributes: {
            url: flowerJPEG,
          },
        },
      ],
    },
  },
};

export const SEARCHBAR_ITEM_DEFAULT_PROPS: SearchbarItemProps = {
  item: {
    slug: 'lovely-flower',
    smallDescription: 'For you, for me, for us',
    name: 'Lovely flower',
    images: {
      data: [
        {
          attributes: {
            url: flowerJPEG,
          },
        },
      ],
    },
  },
  handleClick: () => void 1,
};

export const SELECT_ITEM = {
  label: 'Label1',
  value: 'value1',
};

export const SELECT_DEFAULT_PROPS: Required<
  Omit<SelectProps, 'placeholder' | 'multiple'>
> = {
  items: [SELECT_ITEM],
  initial: ['value1'],
  onChange: () => console.log('changed'),
};

export const PRODUCT: ProductSidebarProps['product'] = {
  slug: 'lovely-flower',
  price: 65,
  name: 'Lovely flower',
  images: {
    data: [
      {
        attributes: {
          url: flowerJPEG,
        },
      },
      {
        attributes: {
          url: roomJPEG,
        },
      },
    ],
  },
  smallDescription: 'For you, for me, for us',
  description: 'Lovely description',
  information: [
    {
      title: 'Title1',
      content: 'content1',
    },
  ],
};

export const PRODUCT_SIDEBAR_DEFAULT_PROPS: ProductSidebarProps = {
  product: PRODUCT,
  handleCartClick: () => {
    console.log('cart');
  },
  handleRemoveCartClick: () => {
    console.log('remove cart');
  },
  handleFavoriteClick: () => {
    console.log('favorite');
  },
  cartValue: new Map([]),
  favoritesValue: new Set('lovely-flower'),
};

export const PRODUCT_INFORMATION_DEFAULT_PROPS: ProductInformationProps = {
  information: [...Array(10).keys()].map((_value, index) => ({
    title: `Title${index + 1}`,
    content: `content${index + 1}`,
  })),
};

export const PRODUCT_GALLERY_DEFAULT_PROPS: ProductGalleryProps = {
  images: [
    {
      attributes: {
        url: flowerJPEG,
      },
    },
    {
      attributes: {
        url: roomJPEG,
      },
    },
  ],
};

const PRODUCT_ORDER_TABLE: OrdersTableProps['orders'][number]['items'][number]['product'] =
  {
    name: 'Flower',
    slug: 'slug-flower',
    description: '',
    smallDescription: '',
    price: 50,
  };

export const ORDERS_TABLE_DEFAULT_PROPS: OrdersTableProps = {
  orders: [
    {
      id: 'id-1',
      amount: 50,
      items: [{ id: '1', amount: 1, product: PRODUCT_ORDER_TABLE }],
    },
    {
      id: 'id-2',
      amount: 100,
      items: [{ id: '2', amount: 2, product: PRODUCT_ORDER_TABLE }],
    },
  ],
};

export const CART_SIDEBAR_DEFAULT_PROPS: CartSidebarProps = {
  cartTotal: 100,
  isUser: true,
  handleCheckoutClick: async () => {
    console.log('checkout');
  },
  isLoading: false,
};

export const CART_ITEM_DEFAULT_PROPS: CartItemProps = {
  inCartAmount: 1,
  item: PRODUCT,
  handleCartMinusClick: () => {
    console.log('cart minus');
  },
  handleCartPlusClick: () => {
    console.log('cart plus');
  },
  handleRemoveFromCart: () => {
    console.log('cart remove');
  },
  toggleFavorite: () => {
    console.log('favorite');
  },
};
