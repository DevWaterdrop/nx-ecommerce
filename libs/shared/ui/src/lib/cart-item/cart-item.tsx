import { ProductCartQuery } from '@nx-ecommerce/shared/graphql/types';
import { createItemHref } from '@nx-ecommerce/shared/utils/create-item-href';
import { formatPrice } from '@nx-ecommerce/shared/utils/format-price';
import clsx from 'clsx';
import { useMemo } from 'react';
import { AmountControl } from './local-components/amount-control';
import { BaseControl } from './local-components/base-control';
import { CoverImage } from './local-components/cover-image';
import { Name } from './local-components/name';

export interface CartItemProps {
  item: NonNullable<
    NonNullable<ProductCartQuery['products']>['data'][number]['attributes']
  >;
  isFavorite?: boolean;
  inCartAmount?: number;
  handleCartMinusClick: () => void;
  handleCartPlusClick: () => void;
  handleRemoveFromCart: () => void;
  toggleFavorite: () => void;
}

export const CartItem: React.FC<CartItemProps> = (props) => {
  const {
    item,
    isFavorite = false,
    inCartAmount = 0,
    handleCartMinusClick,
    handleCartPlusClick,
    handleRemoveFromCart,
    toggleFavorite,
  } = props;

  const { slug, price, name, images } = item;

  const href = useMemo(() => createItemHref(slug), [slug]);

  return (
    <div className={clsx('w-full', 'flex')}>
      <CoverImage href={href} images={images} name={name} />
      <div className={clsx('ml-8 w-full', 'flex flex-col')}>
        <div
          className={clsx(
            'w-full',
            'flex flex-wrap items-center justify-between gap-x-1'
          )}
        >
          <Name name={name} href={href} />
          <span className="font-bold">{formatPrice(price * inCartAmount)}</span>
        </div>
        <small className="mt-1">{item.smallDescription}</small>
        <div
          className={clsx(
            'mt-auto',
            'flex gap-8 items-center flex-wrap',
            'md:(gap-4)',
            'lg:(gap-8)'
          )}
        >
          <AmountControl
            inCartAmount={inCartAmount}
            handleCartMinusClick={handleCartMinusClick}
            handleCartPlusClick={handleCartPlusClick}
          />
          <BaseControl
            isFavorite={isFavorite}
            handleRemoveFromCart={handleRemoveFromCart}
            toggleFavorite={toggleFavorite}
          />
        </div>
      </div>
    </div>
  );
};
