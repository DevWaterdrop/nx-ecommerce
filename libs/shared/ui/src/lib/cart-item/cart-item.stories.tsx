import { Story, Meta } from '@storybook/react';
import { CART_ITEM_DEFAULT_PROPS } from '../../props/constants';
import { CartItem, CartItemProps } from './cart-item';

export default {
  component: CartItem,
  title: 'Components/Cart/Item',
} as Meta;

const Template: Story<CartItemProps> = (args) => {
  return (
    <div className="max-w-[800px]">
      <CartItem {...args} />
    </div>
  );
};

export const Primary = Template.bind({});
Primary.args = {
  ...CART_ITEM_DEFAULT_PROPS,
};
