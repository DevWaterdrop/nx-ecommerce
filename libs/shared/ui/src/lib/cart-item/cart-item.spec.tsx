import { render, RenderResult, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';

import { CartItem } from './cart-item';
import { CART_ITEM_DEFAULT_PROPS } from '../../props/constants';
import { formatPrice } from '@nx-ecommerce/shared/utils/format-price';

describe('CartItem', () => {
  let renderResult: RenderResult;

  beforeEach(() => {
    renderResult = render(<CartItem {...CART_ITEM_DEFAULT_PROPS} />);
  });

  it('should render successfully', () => {
    const { baseElement } = renderResult;

    const tree = renderer
      .create(<CartItem {...CART_ITEM_DEFAULT_PROPS} />)
      .toJSON();
    expect(tree).toMatchSnapshot();

    expect(baseElement).toBeTruthy();
  });

  it('main elements are exists and properly formatted', () => {
    const { item, inCartAmount } = CART_ITEM_DEFAULT_PROPS;
    const { price, name, smallDescription } = item;

    const properPrice = formatPrice(price * (inCartAmount || 0));

    screen.getByText(name);
    screen.getByText(properPrice);
    screen.getByText(smallDescription);
  });
});
