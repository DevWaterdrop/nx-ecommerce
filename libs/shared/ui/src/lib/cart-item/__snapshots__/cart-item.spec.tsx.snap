// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem should render successfully 1`] = `
<div
  className="w-full flex"
>
  <a
    href="/i/lovely-flower"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <div
      className="relative aspect-square w-28 object-cover bg-gray-100"
    >
      Next image stub
    </div>
  </a>
  <div
    className="ml-8 w-full flex flex-col"
  >
    <div
      className="w-full flex flex-wrap items-center justify-between gap-x-1"
    >
      <a
        className="text-base font-semibold uppercase hover:(underline)"
        href="/i/lovely-flower"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Lovely flower
      </a>
      <span
        className="font-bold"
      >
        65.00$
      </span>
    </div>
    <small
      className="mt-1"
    >
      For you, for me, for us
    </small>
    <div
      className="mt-auto flex gap-8 items-center flex-wrap md:(gap-4) lg:(gap-8)"
    >
      <div
        className="h-8 flex"
      >
        <button
          className="h-full w-8 relative group flex items-center justify-center border rounded-l-xl transition hover:(bg-black text-white border-black)"
          onClick={[Function]}
          title="remove 1 item"
          type="button"
        >
          <div
            aria-hidden="true"
            className="h-1/2 w-1/2 relative group-hover:(filter invert brightness-0)"
          >
            Next image stub
          </div>
        </button>
        <span
          className="w-8 flex items-center justify-center border-t border-b"
        >
          1
        </span>
        <button
          className="h-full w-8 relative group flex items-center justify-center border rounded-r-xl transition hover:(bg-black text-white border-black)"
          onClick={[Function]}
          title="add 1 item"
          type="button"
        >
          <div
            aria-hidden="true"
            className="h-1/2 w-1/2 relative group-hover:(filter invert brightness-0)"
          >
            Next image stub
          </div>
        </button>
      </div>
      <div
        className="gap-8 items-center hidden text-xs xs:(flex) sm:(text-base) md:(text-xs gap-4) lg:(text-base gap-8)"
      >
        <button
          className="hover:(underline)"
          onClick={[Function]}
          type="button"
        >
          Remove item
        </button>
        <button
          className="hover:(underline)"
          onClick={[Function]}
          type="button"
        >
          Add to favorites
        </button>
      </div>
    </div>
  </div>
</div>
`;
